### ISSUE TITLE
Generic for named type

### ISSUE THREAD

**Issue by PXshadow** [go2hx/go2hx#101]
```go
type S[T any] struct {
	J T
}
func (d Data[T]) run() {

}
```

Needs to be supported however the same restrictions that caused all generic functions to become macro functions also applies to a type generic. The main issues being T being a type union of number types with operator overloads makes it not possible to have the underlying type haxe.Int64 or Int or haxe.UInt64. Working on a way to support this if no more sophisticated way can be found I'll just have go2hx generate all of the possible generic types from the type union, though I really don't want to do that.

**Comment by elliott5** [comment]
Presumably `T any` could also include any user-defined or Go-standard-package-defined types too, in addition to the Go built-in types. 

That is a very long list indeed, that could only be known after all of the packages comprising a program had been compiled together and a combined list generated.

This would mean that all of the Go packages required in a go2hx library for Haxe would need to be transpiled in the same run.

It is a shame that the native Haxe generic constructs can't be used.

**Comment by ShaharMS** [comment]
For once, i think i have a viable idea:

You can have a file named `TCombos.hx` at `src/_internal` (to hide it from intellisense), which is going to contains all of the needed abstract definitions:

Lets say you need a generic of type Float, Int and Int64, or any 3 types, you add this to `TCombos.hx`
```haxe
abstract OneOfThree<T1, T2, T3>(Dynamic) from T2 from T2 from T3 to T1 to T2 to T3 {}
```
Then, wherever you need an input of one of those 3 types, you can do:

```haxe
import _internal.TCombos;

function iNeed3Types(type:OneOfThree<Int, Int64, Float>)

class SomeClass<T:OneOfThree<Int, Int64, Float>> {}
```

**NOTICE** - the underlying type is dynamic so you probably want to convert it to something else

**Comment by PXshadow** [comment]
Also I just noticed I'm pretty sure the T generic with the OneOfThree constraint type is redundant, since whatever T is can always be replaced with ``OneOfThree<Int, Int64, Float>``
```haxe
class SomeClass<T:OneOfThree<Int, Int64, Float>> {}
```

**Comment by PXshadow** [comment]
```haxe
import stdgo.StdGoTypes;

function main():Void {
	stdgo.fmt.Fmt.println(_m(((10 : GoInt))));
	stdgo.fmt.Fmt.println(_m(((10.2 : GoFloat64))));
}

function _m<T:AnyOf2<GoInt, GoFloat64>>(_x:T):T {
	return _x + _x;
}

@:transitive @:anyof extern abstract AnyOf2<T0, T1>(Dynamic) from T0 from T1 to T0 to T1 {
	@:op(A + B)
	static function add(a:AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>,
		b:AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>):AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>;
}
```

``Cannot add _m.T and _m.T``

However this version does work without ``T`` , ``T`` is needed though for Macro Type inspection of the generic at compile time.

```haxe
package github_com.go2hx.go4hx.rnd;

import stdgo.StdGoTypes;

function main():Void {
	stdgo.fmt.Fmt.println(_m(((10 : GoInt))));
	stdgo.fmt.Fmt.println(_m(((10.2 : GoFloat64))));
}

function _m(_x:AnyOf2<GoInt, GoFloat64>):AnyOf2<GoInt, GoFloat64> {
	return _x + _x;
}

@:transitive @:anyof extern abstract AnyOf2<T0, T1>(Dynamic) from T0 from T1 to T0 to T1 {
	@:op(A + B)
	static function add(a:AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>,
		b:AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>):AnyOf2<stdgo.StdGoTypes.GoInt, stdgo.StdGoTypes.GoFloat64>;
}
```

@haxiomic do you have any ideas about why the T type in the first example doesn't use the add op function?

**Comment by PXshadow** [comment]
Also another issue is Generic types of a Union needing a default type somehow even though it could potentially be any of the types of the Union.

**Comment by elliott5** [comment]
That the Haxe compiler's generation of instantiated code from generic functions does not know how to add two values together using the instantiated type is a severe limitation.

**Comment by PXshadow** [comment]
> That the Haxe compiler's generation of instantiated code from generic functions does not know how to add two values together using the instantiated type is a severe limitation.

Yes indeed @elliott5 it's quite frustrating, generics may be the least supported aspect compared to Golang that go2hx should have 1st class support for.

Perhaps yor initial idea of just generating all potential generic types might be right after all, the limitations are very hard to get around with still using generics though I will try to explore the option a little longer.

**Comment by elliott5** [comment]
The options for implementing Go generics in go2hx would seem to be:

1. Use Haxe generics ... but as the discussion above demonstrates, this is difficult to achieve because of the mis-match between the two language models;
2. Use brute force expansion to generate specific code for every possible Go type that could use the generic ... but this requires a two-pass approach (the first to gather all the types) and so does not permit individual module/package  compilation, plus it will generate a huge amount of code which will slow compilation; or
3. Use an interpretive approach to implementing Go generics, passing and returning all values as Go `interface{}`s ( for a working example of this methodology see the "add" function of Yaegi - https://github.com/traefik/yaegi/blob/2248851d7700bee036109be660b5fbaef1d7b095/interp/op.go#L13 ) ... this is slower at runtime, but simpler to implement.

**Comment by PXshadow** [comment]
> Of course, the best result would be if the project could use Haxe generics.

* Go Generic function = Macro function
* Go Generic type = Haxe Generic type
* Go Generic recv function = [Static extension + Macro function](https://haxe.org/manual/macro-limitations-static-extension.html)

To fix Haxe generic type having null generic type fields, I have it so that when the struct type gets created if it has type params / is a generic type, it will set all struct fields to the default values in the new construction / ``:structInit`` ObjectDeclaration

Struct types will no longer be special with the ability to not have a wrapper or static extension class and now all Named types will have these 2 corresponding classes with them.

Wrapper is redundant to manually type however the macro generation approach before caused many issues, so for the time being this should work out well.

**Comment by elliott5** [comment]
I'm delighted you have found a way forward with this @PXshadow - well done!

**Comment by PXshadow** [comment]
So much macro writing.....

This is the current issue:
``Field index for _add not found on prototype github_com.go2hx.go4hx.rnd.L_static_extension``

It involves referencing a static extension macro but it doesn't exist at compile time and possibly runtime.

**Comment by elliott5** [comment]
Wow @PXshadow, your solution has come a long way!

Does this issue occur when a generic method contains `A + B` but their type is unknown? If so, one approach would be to automatically create methods for all Go types for all Go infix operators (like "+"). Then rely on Haxe DCE to remove them after compilation.

Otherwise, you will need to explain further ;)

**Comment by PXshadow** [comment]
Actions speak louder then words, very pleased with a first draft now committed @elliott5  I wrote up what's still missing in the commit and I hope it's clear enough to see the current state of this. 

As for the issue it occurs because of the strange limbo state of static extension (or Go receiver function) macro functions. 

The function exists at compile time to utilize adding methods to any type with the ``:using`` system but does not exist at runtime...  because macro functions become the returned expression, which is fine if you directly call the function however if the function instead gets stored it becomes a lot more tricky.

The very very good news though is that this potential issue can be overcome by calling a macro function on the static extension macro to create a real function for it. 

An example when this situation would occur would be for passing around the function in this example:

```haxe
package main

func main() {
	l := L[string]{"hello"}
	x := l.add // not called directly but rather stored
	println(x(l)) // now called here
}

type L[T int | string] struct {
	x T
}

func (x L[T]) add(a L[T]) T {
	return x.x + a.x
}
```

I can give the equivalent generated code in Haxe with notes, it is rather long, if you're interested let me know.

Thanks for your patience @elliott5 hopefully this will be the last multi week retreat for implementing major language features very incompatible with Haxe.

**Comment by elliott5** [comment]
Well done for committing a first draft @PXshadow.

If I have understood correctly, because go2hx is using Haxe macros to implement the methods (receiver functions) of a Go generic type, that code must be generated at compile time. 

Thus when a Go generic type is instantiated, all of its methods must also be created. We'll need all of them because if the instance is passed as a Go `interface` any of those methods could be called.

My suggestion is to auto-generate both the additional fields within the generic type to hold the methods, and the code to populate those additional fields.

To slightly re-write your example:
```
package main

func main() {
	l := L[string]{
		"hello",
		func(x, a L[string]) string { return x.addMacro(a) }, // auto generated line
	}
	println(l.add(l, l)) // auto changed call signature
}

type L[T int | string] struct {
	x   T
	add func(x L[T], a L[T]) T // auto generated line
}

func (x L[T]) addMacro(a L[T]) T { // auto generated re-name
	return x.x + a.x
}
```

I'm hoping there are more elegant ways of implementing the slightly convoluted Go above once transpiled into Haxe!

Since many generic variables of the same implementation type will be created, possibly repeatedly, there will be more efficient ways of implementing the above by holding the instantiated type methods in a per-generic registry, and generating them only once per program execution... but that is for the future.

**Comment by PXshadow** [comment]
@elliott5 Yes you are correct. Basically macro functions must be called straight away so in order to have the Haxe compiler not complain.

There is a Haxe compiler issue holding back the 2nd draft from fully working, hopefully it will be addressed soon https://github.com/HaxeFoundation/haxe/issues/10716

**Comment by elliott5** [comment]
As it may take some time for the Haxe compiler team to resolve this issue @PXshadow, is there some naming convention you could adopt that would make the issue less likely to occur?

**Comment by PXshadow** [comment]
@elliott5 yes even better I could check for name conflicts and then simply make those functions non modular and instead in a class, shouldn't anything functionality wise, only slightly less good looking syntax output.

**Comment by elliott5** [comment]
I see @Simn has already fixed https://github.com/HaxeFoundation/haxe/issues/10716 ... thank you Simon, the Haxe foundation is very lucky to employ you!

**Comment by PXshadow** [comment]
I assume this is fixed now, feel free to write if not and I will reopen.
